joomla/joomla-cms
# Add locked field to extensions table, prevent uninstalling core extensions, restructure protected extension list

Pull Request for Issue #13037

### Summary of Changes

This is PR #13037 updated to the current 3.9 branch.  See the original PR for added context and past conversation.

The long and short is this adds a new field, `locked`, to the extensions table, and splits the current definition of the `protected` field.  This also updates the protected extension list to make the list only include admin components, core libraries, and the files extension that Joomla tracks itself with (this means all modules and plugins are now unprotected and can be enabled/disabled consistently).

Current:
- `protected` indicates an extension which cannot be disabled or uninstalled

New:
- `protected` indicates an extension which cannot be disabled
- `locked` indicates an extension which cannot be uninstalled

#### Why This Distinction Matters

Uninstalling core extensions can be problematic, and honestly it's not very effective given our current packaging and upgrading solutions.  Even if you do uninstall the extensions, they end up back on your site's filesystem during the update process because we don't make the package extraction step aware of the database and inherently the installed extensions.  Also, uninstalling components takes their tables with them, and if an update includes a SQL delta for one of those tables, it causes the update to fail over.  So we should take an extra step to protect users from doing things that could be dangerous for their sites.  Next, an uninstallable extension should not mean that the extension must be enabled.  The only extensions which should be protected are those which if disabled would critically bring down the site, every other aspect of the extension listing should be controllable by the site administrator.

### Testing Instructions

With the patch applied and relevant SQL updates applied, the user should be unable to uninstall any core extension (on a correct install this is any extension with an ID less than 10000).  The user should not be able to disable extensions which are critical to the management of the site (i.e. the extension manager, plugin manager, or update component, any of the libraries, or Joomla files extension) or are coupled to the core application stack (i.e. com_categories or com_content), all other extensions should be able to be disabled through the Extension Manager interface.

### Documentation Changes Required

We should probably add a page explaining the various extension states and what these columns exist for if one doesn't already exist.
